### PR TITLE
Enable StructEnreg by default on all platforms.

### PR DESCRIPTION
**Pull request by sandreenko**
Diffs are positive and looks similar to x64 diffs merged in https://github.com/dotnet/runtime/pull/55045 and x86 merged in https://github.com/dotnet/runtime/pull/55535.

```
coreclr_tests.pmi.Linux.arm64.checked.mch
Total bytes of delta: -3112
153 total files with Code Size differences (151 improved, 2 regressed), 7 unchanged.

libraries.crossgen2.Linux.arm64.checked.mch
Total bytes of delta: -4460
561 total files with Code Size differences (462 improved, 99 regressed), 49 unchanged.

libraries.pmi.Linux.arm64.checked.mch
Total bytes of delta: -5300
111 total files with Code Size differences (95 improved, 16 regressed), 11 unchanged.

libraries_tests.pmi.Linux.arm64.checked.mch
Total bytes of delta: -324
85 total files with Code Size differences (58 improved, 27 regressed), 17 unchanged.

benchmarks.run.Linux.x64.checked.mch
Total bytes of delta: -390
5 total files with Code Size differences (5 improved, 0 regressed), 0 unchanged.

coreclr_tests.pmi.Linux.x64.checked.mch
Total bytes of delta: -4996
157 total files with Code Size differences (156 improved, 1 regressed), 4 unchanged.

libraries.crossgen2.Linux.x64.checked.mch
Total bytes of delta: -3153
271 total files with Code Size differences (224 improved, 47 regressed), 18 unchanged.

libraries.pmi.Linux.x64.checked.mch
Total bytes of delta: -970
80 total files with Code Size differences (68 improved, 12 regressed), 0 unchanged.

Total bytes of delta: -997
libraries_tests.pmi.Linux.x64.checked.mch
85 total files with Code Size differences (69 improved, 16 regressed), 4 unchanged.

benchmarks.run.windows.arm64.checked.mch
Total bytes of delta: -160
5 total files with Code Size differences (5 improved, 0 regressed), 0 unchanged.

coreclr_tests.pmi.windows.arm64.checked.mch
Total bytes of delta: -3120
153 total files with Code Size differences (151 improved, 2 regressed), 7 unchanged.

libraries.crossgen2.windows.arm64.checked.mch
Total bytes of delta: -4536
572 total files with Code Size differences (472 improved, 100 regressed), 48 unchanged.

libraries.pmi.windows.arm64.checked.mch
Total bytes of delta: -5376 (-5.36% of base)
118 total files with Code Size differences (102 improved, 16 regressed), 11 unchanged.

libraries_tests.pmi.windows.arm64.checked.mch
Total bytes of delta: -308
85 total files with Code Size differences (56 improved, 29 regressed), 18 unchanged.

coreclr_tests.pmi.Linux.arm.checked.mch
Total bytes of delta: -1660
49 total files with Code Size differences (48 improved, 1 regressed), 0 unchanged.

libraries.crossgen2.Linux.arm.checked.mch
Total bytes of delta: -1280
208 total files with Code Size differences (185 improved, 23 regressed), 13 unchanged.

libraries.pmi.Linux.arm.checked.mch
Total bytes of delta: -34
3 total methods with Code Size differences (2 improved, 1 regressed), 0 unchanged.

libraries_tests.pmi.Linux.arm.checked.mch
Total bytes of delta: -118
7 total files with Code Size differences (7 improved, 0 regressed), 4 unchanged.
```

The regressions are the same: 
1. we use different frame size and it could require bigger init code;
2. we use more registers and have to store/restore non-volatile.

The fixes that are included in this PR:
1. `genBitCast` on arm64/arm32 was not working correctly with regOptional lclVars, did not repro without additional stress pressure, the bug was that if a local was not assigned a register and we had a bitcast on top of it we were marking it as contained and were trying to load directly but for locals the direct load happens in `genConsumeRegs` that has already happened there and it was skipped because the local was contained;
2. arm32 prespill all struct args, so they are excluded from register allocation;
3. SIMDIntrinsicUpperSave/Restore for SIMD regs when calling conv does not keep them, so if we enregister struct<16> as `v8` we have to do bookkeeping for the upper part around each call etc.

the problem with these fixes was that they all were silent bad codegen, I have covered the 3 with new asserts but others are too low in the pipeline to be checked.

Contributes to https://github.com/dotnet/runtime/issues/43867 and probably closes the scope for 6.0.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

**Comment by kunalspathak** [comment]
Does this fix ARM64 issue https://github.com/dotnet/runtime/issues/35071?

File: src/coreclr/jit/lclvars.cpp [hunk 1]
```diff
@@ -3493,13 +3493,42 @@ void Compiler::lvaSortByRefCount()
...
+                {
+                    lvaSetVarDoNotEnregister(lclNum DEBUGARG(DNER_IsStruct));
+                }
+                else if (varDsc->lvIsMultiRegArgOrRet())
+                {
+                    // Prolog and return generators do not support SIMD<->general register moves.
+                    lvaSetVarDoNotEnregister(lclNum DEBUGARG(DNER_IsStructArg));
+                }
+            }
+            else if (varDsc->lvType == TYP_STRUCT)
```
Comment (BruceForstall) [inline src/coreclr/jit/lclvars.cpp:3512]: This is the same condition as before; this block appears to be dead code.

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/lclvars.cpp (hunk 1 above)
Comment (sandreenko) [reply src/coreclr/jit/lclvars.cpp:3512]: looks like a merge issue and that is why I am seeing some failures in CI, will fix it soon.

Comment (sandreenko) [reply src/coreclr/jit/lclvars.cpp]: fixed, thank you.

**Review by BruceForstall — APPROVED** [review]

**Comment by sandreenko** [comment]
The UI here does not show the test result (yeah, ci) but:
1) [runtime ](https://dev.azure.com/dnceng/public/_build/results?buildId=1238988&view=results)is green;
2) [runtime-jitstress](https://dev.azure.com/dnceng/public/_build/results?buildId=1239052&view=ms.vss-test-web.build-test-results-tab) is green
3) [runtime jitstressregs](https://dev.azure.com/dnceng/public/_build/results?buildId=1239053&view=ms.vss-test-web.build-test-results-tab&runId=36861852&resultId=109775&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab) only one unrelated failure, covered by https://github.com/dotnet/runtime/issues/11063#issuecomment-880272558
4) [runtime jitstress2-jitstressregs](https://dev.azure.com/dnceng/public/_build/results?buildId=1239054&view=ms.vss-test-web.build-test-results-tab) shows one unrelated failure covered by https://github.com/dotnet/runtime/issues/55449#issuecomment-880412252;

so the tests are passing, I will keep an eye on library stress tests but we have time to fix bugs if anything pops up! Thank Bruce again for the quick review.

**Pull request merged**